File: js/examples/insurance-claim/functions/approveClaim.ts
import { createSomaFunction } from '@soma/sdk/bridge';
import type { ProviderController } from '@soma/sdk';
import { assessmentSchema } from '../agents';
import z from 'zod';
export const providerController: ProviderController = {
    typeId: 'approve-claim',
    name: 'Approve Claim',
    documentation: 'Approve a claim',
    categories: [],
    credentialControllers: [
        // {
        //     type: "Oauth2AuthorizationCodeFlow",
        //     field0: {
        //         staticCredentialConfiguration: {
        //             authUri: 'https://random-number.com/auth',
        //             tokenUri: 'https://random-number.com/token',
        //             userinfoUri: 'https://random-number.com/userinfo',
        //             jwksUri: 'https://random-number.com/jwks',
        //             issuer: 'https://random-number.com',
        //             scopes: ['openid', 'profile', 'email'],
        //         },
        //     }
        // }
        {
            type: "NoAuth",
        }
    ],
};

export default createSomaFunction({
    inputSchema: assessmentSchema,
    outputSchema: z.object({
        approved: z.boolean(),
    }),
    providerController,
    functionController: {
        name: 'approve-claim',
        description: 'Approve a claim',
    },
    handler: async ({claim})=> {
        console.log('Claim', claim);
        // perform an async action here to approve the claim
        return { approved: true };    
    },
});

File: py/examples/insurance-claim-bot/functions/approve_claim.py
"""Function to approve insurance claims."""

from pydantic import BaseModel

from soma_sdk.bridge import (
    ProviderController,
    FunctionController,
    create_soma_function,
)


class InsuranceClaim(BaseModel):
    """Insurance claim details."""

    date: str
    category: str
    reason: str
    amount: float
    email: str


class Assessment(BaseModel):
    """Assessment containing the claim."""

    claim: InsuranceClaim


class ApprovalResult(BaseModel):
    """Result of claim approval."""

    approved: bool


provider_controller = ProviderController(
    type_id="approve-claim",
    name="Approve Claim",
    documentation="Approve a claim",
    categories=[],
    credential_controllers=[
        {"type": "NoAuth"},
    ],
)

function_controller = FunctionController(
    name="approve-claim",
    description="Approve a claim",
)


async def approve_claim_handler(input_data: Assessment) -> ApprovalResult:
    """Handler that approves claims.

    In a real application, this would integrate with your
    claims processing system.

    Args:
        input_data: The assessment containing the claim to approve.

    Returns:
        ApprovalResult indicating whether the claim was approved.
    """
    # In a real app, you would:
    # 1. Validate the claim
    # 2. Check business rules
    # 3. Update your database
    # 4. Send notifications
    print(f"Approving claim: {input_data.claim}")
    return ApprovalResult(approved=True)


# Export the function
default = create_soma_function(
    input_schema=Assessment,
    output_schema=ApprovalResult,
    provider_controller=provider_controller,
    function_controller=function_controller,
    handler=approve_claim_handler,
)
